### PR TITLE
Cohort: run participation

### PR DESCRIPTION
**Pull request by rwaldron**

### CODE REVIEW

File: client/components/Facilitator/Components/Cohorts/Cohort.css [hunk 1]
```diff
@@ -58,3 +73,20 @@
 .cohort__table-cell--title-button {
     background: none;
 }
+
+.cohort__tab-menu--overflow > div {
+    /*    overflow-x: auto;
+    overflow-y: hidden;
+*/
+}
```
Comment (rwaldron) [inline client/components/Facilitator/Components/Cohorts/Cohort.css]: Need to figure out what to do wit "too many tabs" in the cohort view.

File: client/components/Facilitator/Components/Cohorts/CohortDataTable.css [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+.table-wrap {
+    position: relative;
+}
+.cohortdatatable__scroll th,
+.cohortdatatable__scroll td {
+    /*  padding: 5px 10px;
+  vertical-align: top;
+*/
+    /*overflow-y: auto;*/
+    /*max-width: 100px;*/
```
[the remark this answers is not in the document]
Comment (rwaldron) [reply client/components/Facilitator/Components/Cohorts/CohortDataTable.css]: Done

File: client/components/Facilitator/Components/Cohorts/CohortDataTable.jsx [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+import './CohortDataTable.css';
+
+export class CohortDataTable extends React.Component {
+    constructor(props) {
+        super(props);
+
+        this.state = {
+            detail: {
+                open: false,
+                noun: '',
```
[the remark this answers is not in the document]
Comment (rwaldron) [reply client/components/Facilitator/Components/Cohorts/CohortDataTable.jsx]: Done

File: client/components/Facilitator/Components/Cohorts/CohortDataTable.jsx [hunk 4]
```diff
@@ -0,0 +1,355 @@
...
+                                                        name="cell"
+                                                        onClick={onCellClick}
+                                                        style={{
+                                                            cursor: 'pointer'
+                                                        }}
+                                                    >
+                                                        {content.substring(
+                                                            0,
+                                                            100
+                                                        )}
```
[the remark this answers is not in the document]
Comment (rwaldron) [reply client/components/Facilitator/Components/Cohorts/CohortDataTable.jsx]: Fixed

File: client/components/Facilitator/Components/Cohorts/CohortDataTable.jsx [hunk 5]
```diff
@@ -0,0 +1,355 @@
...
+                                    );
+                                })}
+                            </Table.Body>
+                            {/*
+                            <Table.Footer>
+                                <Table.Row>
+                                    <Table.HeaderCell />
+                                    <Table.Cell colSpan={prompts.length}></Table.Cell>
+                                </Table.Row>
+                            </Table.Footer>
```
[the remark this answers is not in the document]
Comment (rwaldron) [reply client/components/Facilitator/Components/Cohorts/CohortDataTable.jsx]: Done

File: client/components/Facilitator/Components/Cohorts/CohortDataTableMenu.jsx [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+        //     user: { username }
+        // } = this.props;
+        // if (!users.find(user => username === user.username)) {
+        //     // For now we'll default all unknown
+        //     // users as "participant".
+        //     await this.props.setCohortUserRole({
+        //         id: Number(id),
+        //         role: 'participant'
+        //     });
+        // }
```
[the remark this answers is not in the document]
Comment (rwaldron) [reply client/components/Facilitator/Components/Cohorts/CohortDataTableMenu.jsx]: Done

File: client/components/Facilitator/Components/Cohorts/CohortParticipants.jsx [hunk 7]
```diff
@@ -106,7 +114,7 @@ export class CohortParticipants extends React.Component {
                     <Table.Header className="cohort__table-thead-tbody-tr">
                         <Table.Row>
                             <Table.HeaderCell colSpan={3}>
-                                Participants{'  '}
+                                Participants ({cohort.users.length}) {'  '}
```
[the remark this answers is not in the document]
Comment (rwaldron) [reply client/components/Facilitator/Components/Cohorts/CohortParticipants.jsx]: Done

File: client/package.json [hunk 8]
```diff
@@ -62,6 +62,7 @@
     "sortablejs": "^1.10.1",
     "style-loader": "^1.0.0",
     "url-loader": "^2.1.0",
+    "uuid": "^3.3.3",
```
Comment (rwaldron) [inline client/package.json]: removed

File: client/reducers/cohort.js [hunk 9]
```diff
@@ -80,22 +83,32 @@ const getCohort = (state = initialState, action) => {
         };
     }
 
+    if (type === GET_COHORT_PARTICIPANTS_SUCCESS) {
+        return {
+            ...state,
+            users: cohort.users
+        };
+    }
```
[the remark this answers is not in the document]
Comment (rwaldron) [reply client/reducers/cohort.js]: Done.

File: client/routes/Routes.jsx [hunk 10]
```diff
@@ -85,12 +90,11 @@ const Routes = () => {
                     )}
                 />
             </ConfirmAuth>
+            */}
```
[the remark this answers is not in the document]
Comment (rwaldron) [reply client/routes/Routes.jsx]: Done

File: server/migrations/sql/20191206214228-cohort-run-v2.sql [hunk 11]
```diff
@@ -0,0 +1,8 @@
+ALTER TABLE cohort_run
+    ADD COLUMN linked_at TIMESTAMPTZ NOT NULL DEFAULT CURRENT_TIMESTAMP;
+-- Up above
+---
+-- Down below
+
+ALTER TABLE cohort_run
+    DROP COLUMN linked_at;
```
[the remark this answers is not in the document]
Comment (rwaldron) [reply server/migrations/sql/20191206214228-cohort-run-v2.sql]: Done

File: server/service/cohort/index.js [hunk 12]
```diff
@@ -26,6 +30,15 @@ router.post('/:id/scenarios', [
...
 router.get('/:id', [requireUser, getCohort]);
+router.get('/:id/scenario/:scenario_id/:user_id', [requireUser, getCohortData]);
+router.get('/:id/scenario/:scenario_id', [requireUser, getCohortData]);
+router.get('/:id/participant/:participant_id', [
+    requireUser,
+    getCohortParticipantData
+]);
+
+// /api/cohort/${id}/run/${runId}
+router.get('/:id/run/:run_id', [requireUserForRun, linkCohort]);
```
Comment (rwaldron) [inline server/service/cohort/index.js]: Done

File: server/service/cohort/db.js [hunk 13]
```diff
@@ -88,6 +88,7 @@ exports.getCohort = async ({ id }) => {
     });
 };
 
+// This is for fetching MY cohorts
 exports.getCohorts = async ({ user_id }) => {
```
[the remark this answers is not in the document]
Comment (rwaldron) [reply server/service/cohort/db.js]: Done

File: client/reducers/cohort.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (ktbee) [reply client/reducers/cohort.js]: Thankya kindly :)

File: client/components/Facilitator/Components/Cohorts/CohortDataTable.jsx [hunk 14]
```diff
@@ -303,16 +303,18 @@ export class CohortDataTable extends React.Component {
...
+                                                                    );
+
+                                                                    const display = isAudioContent ? (
+                                                                        <audio
+                                                                            src={`/api/media/${content}`}
+                                                                            controls="controls"
+                                                                        />
+                                                                    ) : (
+                                                                        content
+                                                                    );
```
Comment (rwaldron) [inline client/components/Facilitator/Components/Cohorts/CohortDataTable.jsx]: Excellent

**Pull request merged**